src/digraph.rs

```diff
@@ -2265,6 +2266,150 @@ impl PyDiGraph {
+        &mut self,
+        py: Python,
+        to_replace: HashSet<usize>,
+        node: usize,
+        force_check_cycle: Option<bool>,
+    ) -> PyResult<()> {
+        fn can_contract(
+            graph: &StablePyGraph<Directed>,
+            nodes: &HashSet<NodeIndex>,
+        ) -> bool {
+            let mut visited = graph.visit_map();
+            let mut visit_next = VecDeque::new();
+
+            // Start with successors of `nodes` that aren't in `nodes` itself.
+            for node in nodes {
+                for edge in graph.edges(*node) {
+                    if !nodes.contains(&edge.target()) {
+                        visit_next.push_back(edge.target());
+                    }
+                }
+            }
+
+            // DFS. If we encounter any of `nodes`, there exists a path from `nodes`
+            // back to `nodes` of length > 1, meaning contraction is disallowed.
+            while let Some(decendant) = visit_next.pop_back() {
+                if visited.is_visited(&decendant) {
+                    continue;
+                }
+
+                for edge in graph.edges(decendant) {
+                    if nodes.contains(&edge.target()) {
+                        // we found a path back to `nodes`
+                        return false;
+                    }
+
+                    visit_next.push_back(edge.target());
+                }
+
+                visited.visit(decendant);
+            }
+
+            true
+        }
+
+        macro_rules! assert_node_exists {
+            ($node_index:ident) => {
+                if self.graph.node_weight($node_index).is_none() {
+                    return Err(PyIndexError::new_err(format!(
+                        "Specified node {} is not in this graph",
+                        $node_index.index()
+                    )));
+                }
+            };
+        }
+
+        if to_replace.is_empty() {
+            return Err(PyValueError::new_err("to_replace must not be empty"));
+        }
+
+        let node_index = NodeIndex::new(node);
+        assert_node_exists!(node_index);
+
+        if self.graph.neighbors_undirected(node_index).next().is_some() {
+            return Err(PyValueError::new_err(
+                "replacement node must not have connections",
+            ));
+        }
+
+        let mut indices_to_remove: HashSet<NodeIndex> =
+            HashSet::with_capacity(to_replace.capacity());
+
+        for node in to_replace {
+            let index = NodeIndex::new(node);
+            assert_node_exists!(index);
+            indices_to_remove.insert(index);
+        }
+
+        if (force_check_cycle == Some(true) || self.check_cycle)
+            && !can_contract(&self.graph, &indices_to_remove)
+        {
+            return Err(DAGWouldCycle::new_err(
+                "Substitution would create cycle(s)",
+            ));
+        }
+
+        let mut pred_to_weights = HashMap::new();
+        let mut succ_to_weights = HashMap::new();
```

Comment:
Wow, I'm surprised/impressed that rustc could implicitly tell this was a `HashMap<NodeIndex, Vec<PyObject>`

Reply:
Yes! It seems like if the usages aren't conflicting, type inference just works. I'm not sure if it's better to be explicit or not. In this case, I dropped the explicit type arguments because `rustfmt` was otherwise wrapping the lines in an ugly way. Let me know if you have a preference.

Reply:
It's fine to leave it out if the rust compiler can figure it out. I agree it's more concise and easy to read this way, I was just surprised that it could since the value is set inside the loop via an entry.